mlflow/models/utils.py

```diff
@@ -481,31 +527,56 @@ def _is_sparse_matrix(x):
                         " which suggests a dictionary input mapping input name to a numpy"
                         " array, but a dict with value type {0} was found.".format(
                             type(pf_input[col_name])
-                        )
+                        ),
+                        error_code=INVALID_PARAMETER_VALUE,
                     )
                 new_pf_input[col_name] = _enforce_tensor_spec(pf_input[col_name], tensor_spec)
         elif isinstance(pf_input, pd.DataFrame):
             new_pf_input = {}
             for col_name, tensor_spec in zip(input_schema.input_names(), input_schema.inputs):
-                new_pf_input[col_name] = _enforce_tensor_spec(
-                    np.array(pf_input[col_name], dtype=tensor_spec.type), tensor_spec
+                pd_series = pf_input[col_name]
+                new_pf_input[col_name] = _reshape_and_cast_pandas_column_values(
+                    col_name, pd_series, tensor_spec
                 )
         else:
             raise MlflowException(
                 "This model contains a tensor-based model signature with input names, which"
-                " suggests a dictionary input mapping input name to tensor, but an input of"
-                " type {0} was found.".format(type(pf_input))
+                " suggests a dictionary input mapping input name to tensor, or a pandas"
+                " dataframe input containing columns mapping input name to list value flattened"
+                " from tensor, but an input of type {0} was found.".format(type(pf_input)),
+                error_code=INVALID_PARAMETER_VALUE,
             )
     else:
+        tensor_spec = input_schema.inputs[0]
         if isinstance(pf_input, pd.DataFrame):
-            new_pf_input = _enforce_tensor_spec(pf_input.to_numpy(), input_schema.inputs[0])
+            num_input_columns = len(pf_input.columns)
+            if num_input_columns == 0:
+                raise MlflowException("Input DataFrame is empty.")
+            elif num_input_columns == 1:
+                new_pf_input = _reshape_and_cast_pandas_column_values(
+                    "0", pf_input[pf_input.columns[0]], tensor_spec
```

Comment:
What does `"0"` mean here?

Reply:
Does `pf_input` always have a column named "0"?

Reply:
`pf_input[pf_input.columns[0]]` select the first column, not selecting column named "0".

Reply:
`The input pandas dataframe column '0'` sounds like the input dataframe has an column named '0'.

Reply:
oh I will update the error message